### PR TITLE
Sync with upstream for Garden - attempt 2

### PR DESCRIPTION
**Pull request by nicholasdeepx**
### Description

Add our prior modifications from the closed PRs to the gz-rendering7 branch from upstream. 
[Link to the task](https://app.clickup.com/t/864eq9ufj).

The library was built as below.

```
# In the main directory:
docker build -t deepx-gz-rendering .
docker run -dt --name test_gz_rendering deepx-gz-rendering:latest bash
docker exec -it test_gz_rendering bash

# In the container, mimic the commands in CI.
cd /root/gazebo/gz-rendering/build
cmake ..
cmake --build . --target install -j 12
ctest
```

### Motivation and Context

To prepare to upgrade to Gazebo Garden.
Many tests might fail because they need display, but we can ignore them.

### How Has This Been Tested?

Through manual build as written earlier, and CI.

### Checklist

- [x] Have you filled the sections above?
- [x] Have you checked to ensure there aren't other open [Pull Requests](../pulls) for the same GitHub Issues/ClickUp Task(s)?
- [x] Have you documented the changes in GitHub Issues and/or ClickUp Task(s) related to this Pull Request?
- [x] Have you added labels, where appropriate, to this Pull Request?

### CODE REVIEW

**Review by wgkassem — COMMENTED** [review]
Is `[update] sync files` the commit that merge-pulled from the upstream?
The rest of the commits are fixes. Is that correct?

**Comment by nicholasdeepx** [comment]
Not quite; this branch starts by checking out the gz-rendering7 branch from upstream. The `[update] sync files` commit contains most of our modifications from the previous PRs. I should have given a clearer commit message 😅

**Comment by nicholasdeepx** [comment]
Note: `ctest` in CI is disabled for now because there are many tests which require display. I have not found a pattern to those tests, and it looks tedious to exclude them one-by-one. Still I think we should try to run locally if we modify this package. The actual test for now might just be at the integration level with Talos.

File: .github/workflows/deepx-ci.yml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        run: |
+          docker build -t deepx-gz-rendering .
+      
+      # Skip ctest for now because many tests are throwing exceptions due to lack of display.
+      - name: Build gz-rendering and test
+        run: |
+          docker run deepx-gz-rendering:latest bash \
+          -c 'cd /root/gazebo/gz-rendering/build; \
+          cmake ..; \
+          cmake --build . --target install;'
```
Comment (wgkassem) [inline .github/workflows/deepx-ci.yml]: Sorry for nitpicking. A blank line before the EOF is missing. Could you please add one? Thank you!

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    libsdformat13=13.5.* \
+    && apt-get -qq -y autoclean \
+    && apt-get -qq -y autoremove \
+    && rm -rf /var/lib/apt/lists/*
+
+WORKDIR "/root/gazebo"
+
+COPY . gz-rendering/
+
+RUN mkdir -p gz-rendering/build
```
Comment (wgkassem) [inline Dockerfile]: blank line before EOF again, sorry...

File: include/gz/rendering/base/BaseScene.hh [hunk 3]
```diff
@@ -809,7 +809,7 @@ namespace gz
       /// \param[in] _name unique object name.
       /// \return Pointer to a text geometry object
       protected: virtual TextPtr CreateTextImpl(unsigned int _id,
-                     const std::string &_name);
+                     const std::string &_name) = 0;
```
Comment (wgkassem) [inline include/gz/rendering/base/BaseScene.hh:812]: Did we add this modification?

File: ogre/src/OgreMarker.cc
Comment (wgkassem) [inline ogre/src/OgreMarker.cc:1]: I guess everything new in this file was added by us?

File: ogre2/include/gz/rendering/ogre2/Ogre2Text.hh [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+      private: friend class OgreScene;
+
+      /// \internal
+      /// \brief Private data pointer
+      private: std::unique_ptr<Ogre2TextPrivate> dataPtr;
+    };
+    }
+  }
+}
+#endif
```
Comment (wgkassem) [inline ogre2/include/gz/rendering/ogre2/Ogre2Text.hh:104]: EOF again, sorry. Also, is this a completely new file? Did we write this?

File: ogre2/src/CMakeLists.txt [hunk 5]
```diff
@@ -87,3 +87,5 @@ gz_build_tests(TYPE UNIT
                ENVIRONMENT GZ_RENDERING_INSTALL_PREFIX=${CMAKE_INSTALL_PREFIX})
 
 install(DIRECTORY "media"  DESTINATION ${GZ_RENDERING_RESOURCE_PATH}/ogre2)
+
+add_subdirectory(media)
```
Comment (wgkassem) [inline ogre2/src/CMakeLists.txt]: I'm really sorry, but EOF here again needs a blank line

File: ogre2/src/media/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1 @@
+add_subdirectory(fonts)
```
Comment (wgkassem) [inline ogre2/src/media/CMakeLists.txt]: Blank line before EOF, here please, sorry

File: ogre2/src/media/fonts/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,4 @@
+file(GLOB files "*ttf" "*png" "*fontdef")
+
+install(FILES ${files} DESTINATION ${IGN_RENDERING_RESOURCE_PATH}/ogre2/media/fonts)
+install(DIRECTORY liberation-sans DESTINATION ${IGN_RENDERING_RESOURCE_PATH}/ogre2/media/fonts)
```
Comment (wgkassem) [inline ogre2/src/media/fonts/CMakeLists.txt]: Blank line before EOF again sorry

File: ogre2/src/media/fonts/ignition-rendering.fontdef [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+  resolution 96
+}
+
+Liberation Sans
+{
+  type truetype
+  source liberation-sans/LiberationSans-Regular.ttf
+  size 18
+  resolution 96
+}
```
Comment (wgkassem) [inline ogre2/src/media/fonts/ignition-rendering.fontdef]: also here sorry

**Review by wgkassem — CHANGES_REQUESTED** [review]
Mostly just missing blank lines before EOFs, and some questions. Sorry for the spam about EOF, it just looks bad when reviewing on GH

File: include/gz/rendering/base/BaseScene.hh (hunk 3 above)
Comment (nicholasdeepx) [reply include/gz/rendering/base/BaseScene.hh:812]: Yes, in [this PR](https://github.com/DeepX-inc/gz-rendering/pull/10/files#diff-20f200b860d03f923ff5c6e3c2badb4cc20d29e9993b00d79a909b5c8b7d8c28).
There used to be `= 0` in ancient times in the origin, but for some reasons it was dropped. Kristjan added it again in the above PR.

File: ogre/src/OgreMarker.cc
Comment (nicholasdeepx) [reply ogre/src/OgreMarker.cc:1]: Yup, originally from [this PR](https://github.com/DeepX-inc/gz-rendering/pull/10/files#diff-296c94cc840062a4f8670d4ca1a11009779ea4026f4e863d8154531e34ef7194).

Comment (nicholasdeepx) [reply ogre/src/OgreMarker.cc:1]: Actually I am not quite sure. Kristjan originally added [that in this PR](https://github.com/DeepX-inc/gz-rendering/pull/10/files#diff-296c94cc840062a4f8670d4ca1a11009779ea4026f4e863d8154531e34ef7194), but I am not quite sure whether he wrote it himself or got it from somewhere else...
(Please ignore my previous comment, now deleted)

File: ogre2/include/gz/rendering/ogre2/Ogre2Text.hh (hunk 4 above)
Comment (nicholasdeepx) [reply ogre2/include/gz/rendering/ogre2/Ogre2Text.hh:104]: This is the same case as the OgreMarker.cc file 😅

**Comment by nicholasdeepx** [comment]
The EoF matter has been addressed 🙇‍♂️

**Review by wgkassem — APPROVED** [review]
We will handle the integration when the PR to Garden is made. Thank you!

**Comment by nicholasdeepx** [comment]
Thank you for the review! I created [a task](https://app.clickup.com/t/863gvzp82) to also fix the ctest issue in deepx-ci.yml.

**Pull request merged**